tool/tctl/common/token_command.go

```diff
@@ -528,7 +579,488 @@ func (c *TokensCommand) List(ctx context.Context, client *authclient.Client) err
 			}
 			return table.AsBuffer().String()
 		}
-		fmt.Fprint(c.stdout, tokensView())
+		fmt.Fprint(c.Stdout, tokensView())
 	}
 	return nil
 }
+
+func pingAuthAndProxy(ctx context.Context, client *authclient.Client, updateGroup string, insecure bool) (*proto.PingResponse, *webclient.PingResponse, error) {
+	// detect proxy address
+	authPong, err := client.Ping(ctx)
+	if err != nil {
+		return nil, nil, trace.Wrap(err, "failed to ping Teleport Auth")
+	}
+
+	proxyAddr := authPong.GetProxyPublicAddr()
+	if proxyAddr == "" {
+		return &authPong, nil, trace.BadParameter("failed to discover Teleport proxy address, make sure the Teleport Proxy service is running with `public_addr` set")
+	}
+	// detect autoupdate version
+	proxyPong, err := webclient.Ping(&webclient.Config{
+		Context:     ctx,
+		ProxyAddr:   proxyAddr,
+		Insecure:    insecure,
+		UpdateGroup: updateGroup,
+	})
+	if err != nil {
+		return &authPong, nil, trace.Wrap(err, "failed to ping Teleport Proxy")
+	}
+	return &authPong, proxyPong, nil
+}
+
+func lookupKubeActiveContext(path string) (string, error) {
+	stdout, stderr, err := runKubectlCommand(path, []string{"config", "current-context"})
+	if err != nil {
+		return "", trace.Wrap(err, "calling kubectl: %s", stderr.String())
+	}
+	kubeContext := strings.TrimSpace(stdout.String())
+	if kubeContext == "" {
+		return "", trace.BadParameter("context not set, and no active kubectl context found")
+	}
+	return kubeContext, nil
+}
+
+func runKubectlCommand(kubectlPath string, args []string) (stdout, stderr bytes.Buffer, err error) {
+	cmd := exec.Cmd{
+		Path:      kubectlPath,
+		Args:      append([]string{kubectlPath}, args...),
+		Env:       os.Environ(),
+		Stdout:    &stdout,
+		Stderr:    &stderr,
+		WaitDelay: 30 * time.Second,
+	}
+	if err := cmd.Run(); err != nil {
+		return stdout, stderr, trace.Wrap(err, "running `%s %s`", kubectlPath, args)
+	}
+
+	return stdout, stderr, nil
+}
+
+func detectOIDC(ctx context.Context, kubectlPath, kubeContext string) (*oidc.DiscoveryConfiguration, error) {
+	kubectlArgs := []string{"--context", kubeContext, "get", "--raw=/.well-known/openid-configuration"}
+	stdout, stderr, err := runKubectlCommand(kubectlPath, kubectlArgs)
+	if err != nil {
+		return nil, trace.Wrap(err, "running kubectl: %s", stderr.String())
+	}
+	var oidcResponse oidc.DiscoveryConfiguration
+
+	err = json.Unmarshal(stdout.Bytes(), &oidcResponse)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to parse oidc response: %s", stdout.String())
+	}
+
+	if oidcResponse.Issuer == "" {
+		return nil, trace.BadParameter("failed to discover OIDC issuer, empty Issuer field")
+	}
+
+	oidcDiscoverCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
```

Comment:
Perhaps a more aggressive timeout here? Or a log message before we try to run the discovery? I am worried that perhaps if this is some internal IP/DNS name, then this will hang and be very confusing for the user.